docs/adr/0008-freeing-objects-and-resources.md

```diff
@@ -0,0 +1,109 @@
+# Freeing objects and resources
+
+* Status: Propsed
+* Deciders: ?
+* Date: 2023-10-10
+
+Prevous discussion: [PR 1394](https://github.com/mozilla/uniffi-rs/issues/1394), [Issue 8](https://github.com/mozilla/uniffi-rs/issues/8)
+
+## Context and Problem Statement
+
+UniFFI must support objects that hold resources like database connections, file handles, network sockets, etc.  This is difficult, because objects like this are handled differently by different languages:
```

Comment:
> Java/Kotlin/Python programmers make a distinction between memory and other types of resources

this sentence clears up ideas in this ADR, you should include it in the context.

Reply:
I also had the same question at first, and wanted to share how I thought through it, in case it's helpful for context or other folks!

IIUC, from the perspective of the foreign code, a UniFFI object is "a pointer to some memory that exists outside of the GC heap". Because it's outside the GC heap, it can't be GCed automatically.

That's why these objects need some kind of destructor, either:

1. An explicit method, like `close()` or `destroy()`, or...
2. A finalizer that's automatically called by the foreign language's runtime. In the JVM, that's `java.lang.Object.finalize()` or `Cleaner`; in Python, that's `__del__`; in Go, that's `runtime.SetFinalizer`, and so on.

On the Rust side, the object might have a handle to a "resource"—like a database connection, a file handle, etc.—or it might not, but that's totally opaque to the foreign code.

In that sense, I don't think UniFFI's `AutoCloseable` machinery today is concerned about "resources" specifically—it's a destructor that tells Rust to drop the object, and free up the memory it uses. Whether or not dropping the object also happens to release "resources" on the Rust side—for example, a rusqlite connection implementing `Drop` to automatically close the connection—is an implementation detail from Kotlin's perspective.